### PR TITLE
trait: Add tests for route trait and read certificate from secrets

### LINKED ISSUE #2574: Read certificates from secrets to setup TLS config for route trait
Currently route trait parameters `tls-certificate`, `tls-key`, `tls-ca-certificate`, `tls-destination-ca-certificate` expects certificates to set as string parameters, but that is a difficult to set as command line parameters, example:
```
-t route.tls-key="-----BEGIN RSA PRIVATE KEY-----MIIEowIBAA..."
```

This is a proposal to read certificates from secrets to setup TLS configuration for route trait.

### PR DESCRIPTION
**Pull request by claudio4j**
* Add parameters to read TLS certificates from secrets
* Add unit tests for route TLS configuration
* Add e2e tests for route TLS configuration
* Add example in route trait documentation

Fix #2574

**Release Note**
```release-note
trait/route: Add support to read certificates from secrets
```

### CODE REVIEW

**Review by squakez — CHANGES_REQUESTED** [review]
I think we can take the opportunity and make this more general than just `secret`. We should use the `path` concept (ie `TLSCertificatePath`) that will read a generic path on the runtime integration. That path will be filled by a secret/configmap/file/volume ... by the user via `--resource` option.

**Review by oscerd — APPROVED** [review]

File: e2e/common/traits/route_test.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+		// Cleanup
+		Expect(Kamel("delete", "--all", "-n", ns).Execute()).Should(BeNil())
+	// })
+	})
+}
+
+func httpsRequest(t *testing.T, url string) string {
+	transCfg := &http.Transport{
+		// ignore self signed SSL certificates
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
```
Comment (astefanutti) [inline e2e/common/traits/route_test.go:90]: It would be better to use the CA that signed the certificate.

File: e2e/common/traits/route_test.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+	. "github.com/apache/camel-k/e2e/support"
+	"github.com/apache/camel-k/pkg/util/openshift"
+)
+
+const(
+	secretName = "my-combined-tls"
+	keyName = "my-key"
+	certName = "my-cert"
+	keyFilePath = "files/key.key"
```
Comment (astefanutti) [inline e2e/common/traits/route_test.go:47]: To improve readability and maintainability, it'd be better to have those filles code-generated. You can find an example at https://github.com/apache/camel-k/blob/69f0a4bc95f4f1efa6458d241fbc503cb979a7b0/e2e/common/build/maven_ca_secret_test.go.

**Review by astefanutti — COMMENTED** [review]
We need to complete the e2e tests with _re-encrypt_ and _pass-through_ cases. That would help making the configuration coherent.

For the _edge_ case, the configuration is independent from the Camel K runtime, but for the other cases, it may be useful to have a consistent way to inject the certificates into the runtime.

Last but not least, it'd be great that the  _re-encrypt_ case covers the use of a certificate provided by service signing certificate secrets, see https://docs.openshift.com/container-platform/4.8/security/certificates/service-serving-certificate.html#add-service-certificate_service-serving-certificate.

File: e2e/common/traits/route_test.go (hunk 1 above)
Comment (claudio4j) [reply e2e/common/traits/route_test.go]: Done.

**Comment by claudio4j** [comment]
@astefanutti more e2e tests were added, to support 5 cases: no tls, edge, reencrypt, reencrypt with service signing and passthrough
Also, added more examples in docs on how to run integrations with the different route options.

File: e2e/common/traits/route_test.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+		}
+	}()
+	assert.Nil(t, err)
+	buf := new(bytes.Buffer)
+	_, err = buf.ReadFrom(response.Body)
+	assert.Nil(t, err)
+	return buf.String()
+}
+
+func createSecrets(ns string) error {
```
[the remark this answers is not in the document]
Comment (claudio4j) [reply e2e/common/traits/route_test.go]: I didn't add this to test_support.go as this particular route secret is tied to generated self-signed certificate, not so generic at this point, but if we need to make a generic one, that would not be too difficult.

**Comment by claudio4j** [comment]
> I think we can take the opportunity and make this more general than just `secret`. We should use the `path` concept (ie `TLSCertificatePath`) that will read a generic path on the runtime integration. That path will be filled by a secret/configmap/file/volume ... by the user via `--resource` option.

It may not help in this case, because by using `--resource` the certificate files are mounted in the integration pod, but the openshift route for the edge case is created from the operator pod, so the certificate file just doesn't exist in operator pod file system.

File: e2e/common/traits/route_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (squakez) [reply e2e/common/traits/route_test.go]: Yeah, I think it would be cleaner to have a `func NewBinarySecret(ns string, name string, data map[string][]byte) error` method which generically takes in a binary content. Then you use it from your private method with the certificates created for the purpose.

File: e2e/common/traits/route_test.go [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+func TestRunRouteNoTLS(t *testing.T) {
+	WithNewTestNamespace(t, func(ns string) {
+		ocp, err := openshift.IsOpenShift(TestClient())
+		if !ocp {
+			t.Skip("This test requires route object which is available on OpenShift only.")
+			return
+		}
+		assert.Nil(t, err)
+
+		Expect(Kamel("install", "-n", ns).Execute()).To(Succeed())
```
Comment (squakez) [inline e2e/common/traits/route_test.go:75]: In order reduce the execution time, I think it would be wiser to run the test into the same method. You will only install the operator once and reuse the same kit along all tests. I used the approach in the [user config test](https://github.com/apache/camel-k/blob/main/e2e/common/config/config_test.go).

File: e2e/common/traits/route_test.go [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+	})
+
+	return keyCertificatePair {
+		Key: privateKeyPem,
+		Certificate: certPem,
+	}
+}
+
+func domainName() string {
+	consoleroute := Route("openshift-console", "console")()
```
Comment (squakez) [inline e2e/common/traits/route_test.go:415]: Any reason why to use `openshift-console`? Could we use something not bound to any product?

File: e2e/common/traits/route_test.go [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+	return strings.ReplaceAll(consoleroute.Spec.Host, "console-openshift-console.", "")
+}
+
+func getService(ns string, name string) (*corev1.Service, error) {
```
Comment (squakez) [inline e2e/common/traits/route_test.go:419]: If it's a generic method, and is not already present in test_support.go, we better move it there and have it available for anybody willing to use it.

File: e2e/common/traits/route_test.go [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+	svc := corev1.Service{}
+	key := ctrl.ObjectKey{
+		Namespace: ns,
+		Name:      name,
+	}
+	err := TestClient().Get(TestContext, key, &svc)
+	return &svc, err
+}
+
+func addAnnotation(ns string, name string, annotations map[string]string) error {
```
Comment (squakez) [inline e2e/common/traits/route_test.go:429]: Same comment as for `addService` method

File: pkg/trait/route.go [hunk 8]
```diff
@@ -174,19 +196,83 @@ func (t *routeTrait) Apply(e *Environment) error {
 	return nil
 }
 
-func (t *routeTrait) getTLSConfig() *routev1.TLSConfig {
+func (t *routeTrait) getTLSConfig() (*routev1.TLSConfig, error) {
+	// a certificate is a multiline text, but to set it as value in a single line in CLI, the user must escape the new line character as \\n
+	// but in the TLS configuration, the certificates should be a multiline string
+	// then we need to replace the incoming escaped new lines \\n for a real new line \n
+	key := strings.ReplaceAll(t.TLSKey, "\\n", "\n")
```
[the remark this answers is not in the document]
Comment (squakez) [reply pkg/trait/route.go:203]: Probably I did not express my concern properly. I was thinking this part would be better read with an approach like the following:
```
var key string
if t.TLSKey != "" {
  if t.TLSKeySecret != "" {
    // error, both key and keysecret provided
  }
  key = // take value from t.TLSKey
} else if t.TLSKeySecret != "" {
  key = // take value from t.TLSKeySecret
} else {
  // error, missing key/key secret
}
```

**Review by squakez — CHANGES_REQUESTED** [review]
I think there are still room for certain improvements. It would benefit the e2e test performance and maintainability of code.

File: e2e/common/traits/route_test.go (hunk 4 above)
Comment (claudio4j) [reply e2e/common/traits/route_test.go:74]: Done.

File: e2e/common/traits/route_test.go (hunk 6 above)
Comment (claudio4j) [reply e2e/common/traits/route_test.go]: Dropped this `getService` in favor of using `Service` from test_support.go.

File: e2e/common/traits/route_test.go (hunk 5 above)
Comment (claudio4j) [reply e2e/common/traits/route_test.go]: Changed to use `dns/cluster` object instead.

File: pkg/trait/route.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (claudio4j) [reply pkg/trait/route.go:211]: The parameters ending in`secret` have higher priority and are used. I added a documentation about reading keys and certificates from local filesystem or `secret`.

**Comment by claudio4j** [comment]
There are issues running the route tests, related to openshift 3 test cluster (CI) complaining about different base domain names. I am investigating this issue.

**Comment by claudio4j** [comment]
I changed the e2e test to skip TLS certificate validation for client http request when it is unable to determine the dns base domain name (from dns/cluster object). So the route tests are green, but other unrelated tests are failing due to timeouts.

**Comment by claudio4j** [comment]
retest this please

**Comment by claudio4j** [comment]
There is a single test "Route reencrypt (with openshift service CA) https works" failing because the route host is not ready yet, that may occur occasionally, we may increase the timeout before the route object is ready for use, but it is my view the tests are in good shape.

File: e2e/common/traits/route_test.go [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+
+		// =============================
+		// TLS Route Edge
+		// =============================
+
+		t.Run("Route Edge https works", func(t *testing.T) {
+			Expect(Kamel("run", "-n", ns, "files/PlatformHttpServer.java",
+				"-t", "route.tls-termination=edge",
+				"-t", "route.tls-certificate-secret=" + refCert,
+				"-t", "route.tls-key-secret=" + refKey,
```
Comment (astefanutti) [inline e2e/common/traits/route_test.go:119]: Why is a secret is needed with `edge` termination?

File: e2e/common/traits/route_test.go [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+		// =============================
+
+		t.Run("Route passthrough https works", func(t *testing.T) {
+			Expect(Kamel("run", "-n", ns, "files/PlatformHttpServer.java",
+				// the --resource mounts the certificates inside secret as files in the integration pod
+				"--resource", "secret:" + secretName + "@/etc/ssl/" + secretName,
+				// quarkus platform-http uses these two properties to setup the HTTP endpoint with TLS support
+				"-p", "quarkus.http.ssl.certificate.file=/etc/ssl/" + secretName + "/tls.crt",
+				"-p", "quarkus.http.ssl.certificate.key-file=/etc/ssl/" + secretName + "/tls.key",
+				"-t", "route.tls-termination=passthrough",
```
Comment (astefanutti) [inline e2e/common/traits/route_test.go:145]: Can't this be automated by the route trait? The user could be able to provide the secret name and the termination type.

File: e2e/common/traits/route_test.go [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+		// =============================
+
+		t.Run("Route Reencrypt https works", func(t *testing.T) {
+			Expect(Kamel("run", "-n", ns, "files/PlatformHttpServer.java",
+				// the --resource mounts the certificates inside secret as files in the integration pod
+				"--resource", "secret:" + secretName + "@/etc/ssl/" + secretName,
+				// quarkus platform-http uses these two properties to setup the HTTP endpoint with TLS support
+				"-p", "quarkus.http.ssl.certificate.file=/etc/ssl/" + secretName + "/tls.crt",
+				"-p", "quarkus.http.ssl.certificate.key-file=/etc/ssl/" + secretName + "/tls.key",
+				"-t", "route.tls-termination=reencrypt",
```
Comment (astefanutti) [inline e2e/common/traits/route_test.go:173]: Can't this be automated by the route trait? The user could be able to provide the secret name and the termination type.

**Review by astefanutti — CHANGES_REQUESTED** [review]
I'm not sure I understand why it's necessary for users to deal with low level resource mounting and Quarkus properties. Could the user provide only the secret name and the termination type, then the trait would take care of mounting the secret and set the Quarkus properties accordingly?

File: deploy/traits.yaml [hunk 12]
```diff
@@ -820,8 +820,17 @@ traits:
   platform: false
   profiles:
   - OpenShift
-  description: The Route trait can be used to configure the creation of OpenShift
-    routes for the integration.
+  description: 'The Route trait can be used to configure the creation of OpenShift
+    routes for the integration. The certificate and key contents may be stored either
```
Comment (astefanutti) [inline deploy/traits.yaml:824]: I'd suggest: "may be source either from"

File: deploy/traits.yaml [hunk 13]
```diff
@@ -833,19 +842,39 @@ traits:
...
-      to the OpenShift documentation for additional information.
+      to the OpenShift route documentation for additional information.
   - name: tls-certificate
     type: string
-    description: The TLS certificate contents.Refer to the OpenShift documentation
+    description: The TLS certificate contents.Refer to the OpenShift route documentation
       for additional information.
+  - name: tls-certificate-secret
+    type: string
+    description: The TLS certificate contents read from a secret. The format is "secret-name[/key-name]",
```
Comment (astefanutti) [inline deploy/traits.yaml:852]: If I understand it correctly, this is not the content read from a Secret. That is the reference of the secret name and key containing the TLS certificate.

File: deploy/traits.yaml [hunk 14]
```diff
@@ -833,19 +842,39 @@ traits:
...
+      it will be read, otherwise you can set a key name separated with a "/".Refer
+      to the OpenShift route documentation for additional information.
   - name: tls-key
     type: string
-    description: The TLS certificate key contents.Refer to the OpenShift documentation
+    description: The TLS certificate key contents.Refer to the OpenShift route documentation
       for additional information.
+  - name: tls-key-secret
+    type: string
+    description: The TLS certificate key contents read from a secret. The format is
```
Comment (astefanutti) [inline deploy/traits.yaml:862]: dito.

File: deploy/traits.yaml [hunk 15]
```diff
@@ -833,19 +842,39 @@ traits:
...
+      separated with a "/".Refer to the OpenShift route documentation for additional
+      information.
   - name: tls-ca-certificate
     type: string
     description: The TLS cert authority certificate contents.Refer to the OpenShift
-      documentation for additional information.
+      route documentation for additional information.
+  - name: tls-ca-certificate-secret
+    type: string
+    description: The TLS cert authority certificate contents read from a secret. The
```
Comment (astefanutti) [inline deploy/traits.yaml:873]: dito.

File: e2e/common/traits/route_test.go (hunk 9 above)
Comment (claudio4j) [reply e2e/common/traits/route_test.go:119]: To use `edge` termination a certificate is required and it comes from the `secret`.

File: e2e/common/traits/route_test.go (hunk 10 above)
Comment (claudio4j) [reply e2e/common/traits/route_test.go:145]: It just happen that using quarkus http requires these quarkus properties for [SSL setup](https://quarkus.io/guides/http-reference#ssl), whatever http client the user may choose it may require a different set of parameters to setup it. For example,   [Netty SSL](https://github.com/apache/camel-k/blob/main/examples/http/NettySecureServer.java#L18) requires mounting `-resource file:KeyStore.jks@/etc/ssl/keystore.jks `. With that said, it would be difficult to look at every SSL setup to automate in route trait.

File: e2e/common/traits/route_test.go (hunk 11 above)
Comment (claudio4j) [reply e2e/common/traits/route_test.go:173]: Same answer as above.

File: deploy/traits.yaml (hunk 12 above)
Comment (claudio4j) [reply deploy/traits.yaml:824]: suggestion accepted.

**Comment by claudio4j** [comment]
> I'm not sure I understand why it's necessary for users to deal with low level resource mounting and Quarkus properties. Could the user provide only the secret name and the termination type, then the trait would take care of mounting the secret and set the Quarkus properties accordingly?

It just happen that using quarkus http requires these quarkus properties for [SSL setup](https://quarkus.io/guides/http-reference#ssl), whatever http client the user may choose it may require a different set of parameters to setup it. For example,   [Netty SSL](https://github.com/apache/camel-k/blob/main/examples/http/NettySecureServer.java#L18) requires mounting `-resource file:KeyStore.jks@/etc/ssl/keystore.jks ` to setup `sslContextParameters` object. With that said, it would be difficult to look at every SSL setup to automate in route trait.

File: e2e/common/traits/route_test.go (hunk 10 above)
Comment (astefanutti) [reply e2e/common/traits/route_test.go:145]: My point is more: could the trait set these properties automatically?

File: deploy/traits.yaml (hunk 13 above)
Comment (claudio4j) [reply deploy/traits.yaml:852]: The certificate is stored in the `secret`, isn't the same thing for configuration purposes ? The user wants to setup the certificate to be sourced either from a `secret`  or directly from CLI.

File: e2e/common/traits/route_test.go (hunk 9 above)
Comment (astefanutti) [reply e2e/common/traits/route_test.go:119]: My understanding is that a certificate is not require for `edge` termination. The router’s default certificate is used for TLS termination, which is from my experience what users tend to rely on.

File: deploy/traits.yaml (hunk 13 above)
Comment (astefanutti) [reply deploy/traits.yaml:852]: The way it is put, we may understand that the property must be the content of the TLS certificate, while it must be the secret containing the certificate. I find the later a lot more descriptive as to what is expected for the end user.

**Comment by astefanutti** [comment]
> > I'm not sure I understand why it's necessary for users to deal with low level resource mounting and Quarkus properties. Could the user provide only the secret name and the termination type, then the trait would take care of mounting the secret and set the Quarkus properties accordingly?
> 
> It just happen that using quarkus http requires these quarkus properties for [SSL setup](https://quarkus.io/guides/http-reference#ssl), whatever http client the user may choose it may require a different set of parameters to setup it. For example, [Netty SSL](https://github.com/apache/camel-k/blob/main/examples/http/NettySecureServer.java#L18) requires mounting `-resource file:KeyStore.jks@/etc/ssl/keystore.jks ` to setup `sslContextParameters` object. With that said, it would be difficult to look at every SSL setup to automate in route trait.

it's not clear to me why the `quarkus.http.ssl.certificate` properties could not be set automatically.

File: e2e/common/traits/route_test.go (hunk 9 above)
Comment (claudio4j) [reply e2e/common/traits/route_test.go:119]: I see the certificate is not required for `edge`, I can remove it, however as this is an e2e test it was my understanding to have the tests to cover all scenarios. I just tested locally with no certificate for edge and it works. Well, I can remove the certificate for edge ssl setup, if that's the case.

File: e2e/common/traits/route_test.go (hunk 10 above)
Comment (claudio4j) [reply e2e/common/traits/route_test.go:145]: Only if user is using quarkus http. If user uses netty, vertx or other http endpoint, they may require different SSL setup, for example the user may setup the `sslContextParameters` to read certificates from a custom path.

Comment (astefanutti) [reply e2e/common/traits/route_test.go:145]: Could the underlying HTTP endpoint used be inferred from the dependency set?

That would simplify a lot the configuration in that case, at least for Quarkus HTTP, as the mounting of the secret could be done automatically and the corresponding properties set accordingly.

File: e2e/common/traits/route_test.go (hunk 11 above)
Comment (astefanutti) [reply e2e/common/traits/route_test.go:173]: As for `edge` termination, it may be simpler to rely on the router's default certificate, and closer to what a typical user would do.

File: e2e/common/traits/route_test.go (hunk 9 above)
Comment (astefanutti) [reply e2e/common/traits/route_test.go:119]: One option is to keep that test as is, and add another one covering the scenario of `edge` termination with the router's default certificate, so we cover the most frequent scenario as well.

Comment (claudio4j) [reply e2e/common/traits/route_test.go:138]: Added another test route of type `edge` with default certificate.

**Comment by claudio4j** [comment]
For testing I am using Openshift on IBM cloud and I noticed the operator-ingress uses a letsencrypt certificate CA to manage the serving service CA certificates, which are different than the serving certificates provided by a standard openshift 4 installation. They use different secrets to store the CA and certificate.s That said, I could not yet find a standard way to retrieve the certificate in either cloud services, to read them and set in the http client of the e2e tests. So, I disabled the https certificate verification on client side of the e2e route tests. So I ask to accept this tests as it is, and I can work on this specific certificate client side validation later on.

File: e2e/common/traits/route_test.go (hunk 11 above)
Comment (claudio4j) [reply e2e/common/traits/route_test.go:193]: done.

File: e2e/common/traits/route_test.go (hunk 10 above)
Comment (claudio4j) [reply e2e/common/traits/route_test.go:165]: Can this be worked on another separated issue ?

File: e2e/common/traits/route_test.go [hunk 16]
```diff
@@ -0,0 +1,401 @@
...
+			assert.Nil(t, err)
+			assert.Equal(t, "Hello " + code, response)
+			Expect(Kamel("delete", "--all", "-n", ns).Execute()).Should(BeNil())
+		})
+
+		// =============================
+		// TLS Route Reencrypt With Openshift Service Serving CA
+		// =============================
+
+		Eventually(IntegrationPodPhase(ns, integrationName), TestTimeoutMedium).Should(Equal(corev1.PodRunning))
```
Comment (astefanutti) [inline e2e/common/traits/route_test.go:219]: I think this can be removed. This causes the tests to fail.

File: e2e/common/traits/route_test.go [hunk 17]
```diff
@@ -0,0 +1,401 @@
...
+				"-t", "route.tls-key-secret=" + refKey,
+				"-t", "container.port=8443",
+			).Execute()).To(Succeed())
+
+			// should wait a little time before the service is ready
+			time.Sleep(2 * time.Second)
+			Eventually(Service(ns, integrationName)()).ShouldNot(BeNil())
+			annotations := make(map[string]string)
+			annotations["service.beta.openshift.io/serving-cert-secret-name"] = servingCertificateSecret
+			err = addAnnotation(ns, integrationName, annotations)
```
Comment (astefanutti) [inline e2e/common/traits/route_test.go:245]: The annotation should not be added to the Service directly. As it is a resource owned by the Integration, it's overwritten anytime the reconciliation of the Integration happens. One option is to define the annotation on the Integration, and have it transferred on the Service with the _owner_ trait. The problem with this approach is that the annotation will be transferred on all the owned resources. Another option would be to add an `annotations` parameter to the _service_ trait, so that extra annotations can be provided only to the Service.

File: e2e/common/traits/route_test.go (hunk 10 above)
Comment (astefanutti) [reply e2e/common/traits/route_test.go:165]: Sure, these tests are already a great move forwards.

Comment (claudio4j) [reply e2e/common/traits/route_test.go:165]: Create https://github.com/apache/camel-k/issues/2619 to track further investigation about this topic. Thanks for reviewing.

File: e2e/common/traits/route_test.go (hunk 16 above)
Comment (claudio4j) [reply e2e/common/traits/route_test.go:219]: Oh, this got in the wrong place when I refactored into a single test run. It will be fixed in the next push.

File: e2e/common/traits/route_test.go (hunk 17 above)
Comment (claudio4j) [reply e2e/common/traits/route_test.go:245]: I see your point and I agree. As this is just a short lived e2e test, can we have it as it is and open a new issue to track this specific service annotation issue ?

File: e2e/common/traits/route_test.go [hunk 18]
```diff
@@ -0,0 +1,401 @@
...
+		// =============================
+
+		t.Run("Route Edge https works", func(t *testing.T) {
+			Expect(Kamel("run", "-n", ns, "files/PlatformHttpServer.java", "-t", "route.tls-termination=edge").Execute()).To(Succeed())
+			Eventually(IntegrationPodPhase(ns, integrationName), TestTimeoutMedium).Should(Equal(corev1.PodRunning))
+			route := Route(ns, integrationName)
+			Eventually(route, TestTimeoutMedium).ShouldNot(BeNil())
+			// must wait a little time after route is created, before an http request, 
+			// otherwise the route is unavailable and the http request will fail
+			time.Sleep(60 * time.Second)
```
Comment (astefanutti) [inline e2e/common/traits/route_test.go:122]: Probably to be replaced with `waitBeforeHttpRequest`.

File: e2e/common/traits/route_test.go (hunk 17 above)
Comment (astefanutti) [reply e2e/common/traits/route_test.go:245]: On one hand, these tests added-value is that they script the user actions, so they are more than just short lived e2e tests. In that case, users would not add the annotation on the Service, and we still don't have an answer. Plus, I wonder if that may actually cause the instability of the test. On the other hand, I appreciate that this PR has grown larger than anticipated, so to give us the time to find a proper solution, solve that instability issue, and have that PR merged, I think we can disable this test. WDYT?

Comment (claudio4j) [reply e2e/common/traits/route_test.go:245]: Agreed.

File: e2e/common/traits/route_test.go (hunk 18 above)
Comment (claudio4j) [reply e2e/common/traits/route_test.go:122]: Forgot to replace, thanks for reviewing.

**Review by astefanutti — APPROVED** [review]
Great work!

**Comment by astefanutti** [comment]
@claudio4j could you please fix the conflicts in `route_test.go` and regenerate `resources.go`? Sorry about that.

**Comment by claudio4j** [comment]
@astefanutti  rebase done. All checks are green.

**Comment by astefanutti** [comment]
@claudio4j thanks!
